p2p/shim.go

```diff
@@ -0,0 +1,297 @@
+
+	"github.com/gogo/protobuf/proto"
+)
+
+// ============================================================================
+// TODO: Types and business logic below are temporary and will be removed once
+// the legacy p2p stack is removed in favor of the new model.
+//
+// ref: https://github.com/tendermint/tendermint/issues/5670
+// ============================================================================
+
+var _ Reactor = (*ReactorShim)(nil)
+
+type (
+	// ReactorShim defines a generic shim wrapper around a BaseReactor. It is
+	// responsible for wiring up legacy p2p behavior to the new p2p semantics
+	// (e.g. proxying Envelope messages to legacy peers).
+	ReactorShim struct {
+		BaseReactor
+
+		Name         string
+		PeerUpdateCh chan PeerUpdate
+		Channels     map[ChannelID]*ChannelShim
+	}
+
+	// ChannelShim defines a generic shim wrapper around a legacy p2p channel
+	// and the new p2p Channel. It also includes the raw bi-directional Go channels
+	// so we can proxy message delivery.
+	ChannelShim struct {
+		Descriptor *ChannelDescriptor
+		Channel    *Channel
+		InCh       chan Envelope
+		OutCh      chan Envelope
+		PeerErrCh  chan PeerError
+	}
+
+	// ChannelDescriptorShim defines a shim wrapper around a legacy p2p channel
+	// and the proto.Message the new p2p Channel is responsible for handling.
+	// A ChannelDescriptorShim is not contained in ReactorShim, but is rather
+	// used to construct a ReactorShim.
+	ChannelDescriptorShim struct {
+		MsgType    proto.Message
+		Descriptor *ChannelDescriptor
+	}
+)
+
+func NewShim(name string, descriptors map[ChannelID]*ChannelDescriptorShim) *ReactorShim {
+	channels := make(map[ChannelID]*ChannelShim)
+
+	for _, cds := range descriptors {
+		chShim := NewChannelShim(cds, 1)
+		channels[chShim.Channel.ID] = chShim
+	}
+
+	rs := &ReactorShim{
+		Name:         name,
+		PeerUpdateCh: make(chan PeerUpdate),
+		Channels:     channels,
+	}
+
+	rs.BaseReactor = *NewBaseReactor(name, rs)
+
+	return rs
+}
+
+func NewChannelShim(cds *ChannelDescriptorShim, buf uint) *ChannelShim {
+	inCh := make(chan Envelope, buf)
+	outCh := make(chan Envelope, buf)
+	peerErrCh := make(chan PeerError, buf)
+
+	return &ChannelShim{
+		Descriptor: cds.Descriptor,
+		Channel:    NewChannel(ChannelID(cds.Descriptor.ID), cds.MsgType, inCh, outCh, peerErrCh),
+		InCh:       inCh,
+		OutCh:      outCh,
+		PeerErrCh:  peerErrCh,
+	}
+}
+
+// proxyPeerEnvelopes iterates over each p2p Channel and starts a separate
+// go-routine where we listen for outbound envelopes sent during Receive
+// executions (or anything else that may send on the Channel) and proxy them to
+// the coressponding Peer using the To field from the envelope.
+func (rs *ReactorShim) proxyPeerEnvelopes() {
+	if rs.Switch == nil {
+		panic("proxyPeerEnvelopes: reactor shim switch is nil")
+	}
+
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for e := range cs.OutCh {
+				msg := proto.Clone(cs.Channel.messageType)
+				msg.Reset()
```

Comment:
Which constructor? I'd rather be explicit and safe here.

Reply:
`NewChannel`. This is a hot-path, so I think the optimization is worth it, but we can benchmark this as well.

Reply:
Hmmm, so the only reason I even call `Reset` is because we're cloning and then unmarshaling during `Receive`, my intuition was to call `Reset` so we can work with a "new/clean"  instantiation of the proto message.

However, in `proxyPeerEnvelopes` do we even need to call `Reset` -- we marshal into the object shortly after.

Reply:
Ah right. So what we want is basically for `messageType` to be an empty message, so that clone gives us an empty copy and we can unmarshal into that. However, there's no guarantee that whoever created the channel gave an empty message, and by calling `Reset()` in the constructor we make sure that `messageType` is an empty message.